nova_vm/src/ecmascript/builtins/proxy.rs

```diff
@@ -770,14 +773,79 @@ impl InternalMethods for Proxy {
         todo!();
     }
 
+    /// ### [10.5.12 [[Call]] ( thisArgument, argumentsList )](https://tc39.es/ecma262/#sec-proxy-object-internal-methods-and-internal-slots-call-thisargument-argumentslist)
+    ///
+    /// The [[Call]] internal method of a Proxy exotic object O takes
+    /// arguments thisArgument (an ECMAScript language value)
+    /// and argumentsList (a List of ECMAScript language values)
+    /// and returns either a normal completion containing an ECMAScript
+    /// language value or a throw completion.
     fn internal_call(
         self,
-        _agent: &mut Agent,
-        _this_value: Value,
-        _arguments_list: ArgumentsList,
-        _gc: GcScope<'_, '_>,
+        agent: &mut Agent,
+        _: Value,
+        arguments: ArgumentsList,
+        mut gc: GcScope<'_, '_>,
     ) -> JsResult<Value> {
-        todo!()
+        let this_argument = arguments.get(1);
+        let arguments_list = arguments.get(2);
+        // 1. Perform ? ValidateNonRevokedProxy(O).
+        // 2. Let target be O.[[ProxyTarget]].
+        // 3. Let handler be O.[[ProxyHandler]].
+        // 4. Assert: handler is an Object.
+        let NonRevokedProxy {
+            mut target,
+            mut handler,
+        } = validate_non_revoked_proxy(agent, self, gc.nogc())?;
+        // 5. Let trap be ? GetMethod(handler, "apply").
+        let trap = if let TryResult::Continue(trap) = try_get_object_method(
+            agent,
+            handler,
+            BUILTIN_STRING_MEMORY.apply.into(),
+            gc.nogc(),
+        ) {
+            trap?
+        } else {
+            let scoped_handler = handler.scope(agent, gc.nogc());
+            let scoped_target = target.scope(agent, gc.nogc());
+            let trap = get_object_method(
+                agent,
+                handler,
+                BUILTIN_STRING_MEMORY.apply.into(),
+                gc.reborrow(),
+            )?
+            .map(Function::unbind);
+            let gc = gc.nogc();
+            let trap = trap.map(|f| f.bind(gc));
+            handler = scoped_handler.get(agent).bind(gc);
+            target = scoped_target.get(agent).bind(gc);
+            trap
+        };
+        // 6. If trap is undefined, then
+        let Some(trap) = trap else {
+            // a. Return ? Call(target, thisArgument, argumentsList).
+            return call(
+                agent,
+                target.into_value(),
+                this_argument,
+                Some(ArgumentsList(&[arguments_list])),
+                gc,
+            );
+        };
+        // 7. Let argArray be CreateArrayFromList(argumentsList).
+        let arg_array = create_array_from_list(agent, &[arguments_list], gc.nogc());
+        // 8. Return ? Call(trap, handler, « target, thisArgument, argArray »).
+        return call(
+            agent,
+            trap.into_value(),
```

Comment:
But here at least we can use `call_function`

Reply:
Thank you for the comment! I’m encountering the following error:

```rs
error[E0505]: cannot move out of `gc` because it is borrowed
   --> nova_vm/src/ecmascript/builtins/proxy.rs:850:13
    |
791 |         mut gc: GcScope<'_, '_>,
    |         ------ binding `gc` declared here
...
808 |             gc.nogc(),
    |             -- borrow of `gc` occurs here
...
841 |         return call_function(
    |                ------------- borrow later used by call
...
850 |             gc,
    |             ^^ move out of `gc` occurs here
```

Using call works without issues, but with call_function, I get this error. Honestly, I’m not entirely sure why this happens. Do you have any ideas about what might be causing this? :eyes:

Reply:
When doing `call_function(agent, trap, .., gc)` add `.unbind()` to the trap parameter.

This is basically the GC safety system in action, it just doesn't like function interface boundaries :)

Reply:
I was able to get it working by applying the fix in https://github.com/trynova/nova/pull/510/commits/4c31cd728b7b053caadc137791c0ceda70e5361c!

Thank you!!! I haven’t really gone through this part of the code much, so I’ll make sure to read it. Rust is tough…

Reply:
Don't worry, the GC safety system is something I'm only managing to get working right now, and it's pretty awkward even though it does work. This also isn't exactly your run of the mill Rust: This is very much abusing the Rust borrow checker's mechanisms to get static checking of ... well, of rooting and garbage collector safepoints.

I'll probably be writing a more complete contributor's guide to this stuff, and a more generally accessible blog post as well. I'll also give a talk on this at FOSDEM at the turn of the month: https://fosdem.org/2025/schedule/event/fosdem-2025-4394-abusing-reborrowing-for-fun-profit-and-a-safepoint-garbage-collector/

Reply:
Oh, that’s exciting!!!
I’m looking forward to the blog, but do you know if there will be an archive of the talk as well…?

Reply:
Yes, unless something goes badly wrong the video will be available online.